crates/turborepo-lib/src/shim.rs

```diff
@@ -0,0 +1,260 @@
+        let mut first_package_json_dir = None;
+        // We loop through these directories and see if there are workspaces defined in
+        // them, either in the `package.json` or `pnm-workspaces.yml`
+        for dir in potential_roots {
+            if first_package_json_dir.is_none() {
+                first_package_json_dir = Some(dir)
+            }
+
+            let pnpm = PackageManager::Pnpm;
+            let npm = PackageManager::Npm;
+            let is_workspace =
+                pnpm.get_workspace_globs(dir).is_ok() || npm.get_workspace_globs(dir).is_ok();
+
+            if is_workspace {
+                return Ok(Self {
+                    root: dir.to_path_buf(),
+                    mode: RepoMode::MultiPackage,
+                });
+            }
+        }
+
+        // Finally, if we don't detect any workspaces, go to the first `package.json`
+        // and use that in single package mode.
+        let root = first_package_json_dir
+            .ok_or_else(|| {
+                anyhow!(
+                    "Unable to find `{}` or `package.json` in current path",
+                    TURBO_JSON
+                )
+            })?
+            .to_path_buf();
+
+        Ok(Self {
+            root,
+            mode: RepoMode::SinglePackage,
+        })
+    }
+
+    /// Attempts to run correct turbo by finding nearest package.json,
+    /// then finding local turbo installation. If the current binary is the
+    /// local turbo installation, then we run current turbo. Otherwise we
+    /// kick over to the local turbo installation.
+    ///
+    /// # Arguments
+    ///
+    /// * `turbo_state`: state for current execution
+    ///
+    /// returns: Result<i32, Error>
+    fn run_correct_turbo(self, shim_args: ShimArgs) -> Result<Payload> {
+        let local_turbo_path = self.root.join("node_modules").join(".bin").join({
+            #[cfg(windows)]
+            {
+                "turbo.cmd"
+            }
+            #[cfg(not(windows))]
+            {
+                "turbo"
+            }
+        });
+
+        let current_turbo_is_local_turbo = local_turbo_path == current_exe()?;
+        // If the local turbo path doesn't exist or if we are local turbo, then we go
+        // ahead and run the Go code linked in the current binary.
+        if current_turbo_is_local_turbo || !local_turbo_path.try_exists()? {
+            cli::run()
+        } else {
+            // Otherwise we spawn the local turbo process.
+            Ok(Payload::Rust(
+                self.spawn_local_turbo(&local_turbo_path, shim_args),
+            ))
+        }
+    }
+
+    fn spawn_local_turbo(&self, local_turbo_path: &Path, mut shim_args: ShimArgs) -> Result<i32> {
+        let cwd = self
+            .root
+            .to_str()
+            .ok_or_else(|| anyhow!("Root directory path is invalid unicode"))?
+            .to_string();
+
+        let mut raw_args: Vec<_> = vec!["--skip-infer".to_string(), "--cwd".to_string(), cwd];
```

Comment:
Shouldn't we be setting the new process's cwd instead of using the flag since it was added semi-recently?

Reply:
Valid. I was not sure if `--cwd` did something special in Go-land